stolon/src/lib.rs

```diff
@@ -0,0 +1,79 @@
+use std::{net::SocketAddr, sync::Arc};
+
+use tokio_socks::{tcp::Socks5Stream, TargetAddr};
+use tower::{Service, ServiceExt};
+
+pub use zebra_chain::{
+    parameters::Network, serialization::ZcashDeserialize, transaction::Transaction,
+};
+
+pub type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+pub async fn obtain_peer_addr(
+    tor_addr: SocketAddr,
+    network: Network,
+) -> Result<SocketAddr, BoxError> {
+    let seed_domain = match network {
+        Network::Mainnet => "mainnet.seeder.zfnd.org:8233",
+        Network::Testnet => "testnet.seeder.zfnd.org:18233",
+    };
+
+    tracing::debug!(?seed_domain, "using tor resolver to find peers");
+    // this API is pretty awkward, let's upstream some patches.
+    let mut peer_addr = match Socks5Stream::tor_resolve(tor_addr, seed_domain).await? {
+        TargetAddr::Ip(addr) => addr,
+        TargetAddr::Domain { .. } => return Err("resolving a domain gave a domain".into()),
+    };
```

Comment:
`tor_resolve` returns a single IP, rather than multiple ones, and seems to do caching.  I'm not sure whether this is a `tokio-socks` limitation or a Tor limitation.  In any case, it prevents the fanout logic from actually working.

Reply:
**Analysis**

The Tor protocol allows multiple responses to a RESOLVE request:
https://gitweb.torproject.org/torspec.git/tree/tor-spec.txt#n1829

But Tor only sends a single IP address per RESOLVED cell:
https://github.com/torproject/tor/blob/master/src/feature/relay/dns.c#L491

And Tor caches DNS answers:
https://github.com/torproject/tor/blob/master/src/feature/relay/dns.c#L38

**Workaround**

As a workaround, make multiple resolve requests to different exits by modifying `tor_resolve` as follows:
```rust
- let sock = Self::execute_command(proxy, target, Authentication::None, Command::TorResolve).await?;
+ let sock = Self::execute_command(proxy, target, Authentication::Password { random_username, b"" }, Command::TorResolve).await?;
```
https://docs.rs/tokio-socks/0.3.0/src/tokio_socks/tcp.rs.html#99

Using a random username makes Tor use a different circuit for each request.

For details, see *IsolateSOCKSAuth* in: https://2019.www.torproject.org/docs/tor-manual.html.en#SocksPort

**Follow-Up**

`tokio-socks`

Using a new exit should be supported via an argument to `tor_resolve`, and all of the tor functions.

Using tor's default slow exit rotation (10+ minutes) for resolves can impact your anonymity - depending on your particular threat model.

`tor`

tor could return multiple addresses in `RESOLVED` cells, saving load on the network.

Reply:
**Alternate Workaround**

If you don't need to know the IP addresses of the Zcash nodes you're using, you can save time and network load by simply asking Tor to connect to the seeder hostname.

Then the Tor exit will resolve the seeder hostname for you, and connect to one of the nodes supplied by the seeder.

I just realised that this would work when reading this bitcoin ticket:
https://github.com/bitcoin/bitcoin/issues/6808

To get different nodes, you'll need to ask Tor for a separate circuit for each connection, by providing randomised authentication.

(Tor actually returns the remote IP address via the SOCKS5 protocol, but that information is not exposed by tokio-socks.)

Reply:
(You could also get a different circuit by using different seeder addresses, but since there are only 3 or 4 of those, depending on the network, that would limit your fanout.)

Reply:
In the short term, I think your alternative fix (resolving the DNS name directly) seems like a good option.

Reply:
Implemented in daad311324db72a7abf9f5481595e3887b5f5749

Reply:
Looks good!